File: native/sites/contribution_receipt.go
package sites

import (
	"encoding/base64"
	"errors"
	"github.com/JohnnyPBelo/relayloom/native/core"
	"sort"
)

const ContributionReceiptBytes = 8192
const ContributionReceiptLifetimeMS int64 = 30 * 86400000

var receiptFields = []string{"contributorId", "certificateId", "operationId", "target", "proposalCreated", "proposalExpires", "verifiedAt", "created", "expires"}

func receiptError() error { return errors.New("recibo de proposta inválido") }
func receiptBody(value any) (map[string]any, core.PublicIdentity, error) {
	var none core.PublicIdentity
	fields := append([]string{"domain", "owner"}, receiptFields...)
	b, err := object(value, fields...)
	if err != nil || b["domain"] != "relayloom/site-contribution-receipt/1" {
		return nil, none, receiptError()
	}
	card, err := bounded(b["owner"], CertificateBytes)
	if err != nil {
		return nil, none, err
	}
	owner, err := core.DecodePublicIdentity(card)
	if err != nil {
		return nil, none, err
	}
	if !core.ValidAddress(docTextValue(b["contributorId"])) || !core.ValidAddress(docTextValue(b["certificateId"])) || !contributionOperationPattern.MatchString(docTextValue(b["operationId"])) {
		return nil, none, receiptError()
	}
	target, err := contributionTarget(b["target"])
	if err != nil {
		return nil, none, err
	}
	actualOwner, _, _ := ParseAddress(docTextValue(target["site"]))
	if actualOwner != owner.ID {
		return nil, none, receiptError()
	}
	times := map[string]int64{}
	for _, k := range []string{"proposalCreated", "proposalExpires", "verifiedAt", "created", "expires"} {
		n, err := contributionClock(b[k])
		if err != nil {
			return nil, none, err
		}
		times[k] = n
	}
	pc, pe, verified, created, expires := times["proposalCreated"], times["proposalExpires"], times["verifiedAt"], times["created"], times["expires"]
	if pe <= pc || pe-pc > ContributionLifetimeMS || verified >= pe || pc-verified > ContributionClockSkewMS || created != verified || expires <= created || expires-created > ContributionReceiptLifetimeMS {
		return nil, none, receiptError()
	}
	if _, err = bounded(b, ContributionReceiptBytes); err != nil {
		return nil, none, err
	}
	return b, owner, nil
}

// VerifyContributionReceipt authenticates a historical attestation. It grants no
// site edit authority, transport permission or extension of a visitor's grant.
func VerifyContributionReceipt(value any) (map[string]any, error) {
	cert, err := object(value, "body", "id", "signature")
	if err != nil {
		return nil, receiptError()
	}
	body, owner, err := receiptBody(cert["body"])
	if err != nil {
		return nil, err
	}
	data, err := bounded(body, ContributionReceiptBytes)
	if err != nil {
		return nil, err
	}
	id, signature := docTextValue(cert["id"]), docTextValue(cert["signature"])
	if !core.ValidAddress(id) || id != core.Hash(data) || len(signature) != 88 {
		return nil, receiptError()
	}
	bytes, err := base64.StdEncoding.Strict().DecodeString(signature)
	if err != nil || len(bytes) != 64 || base64.StdEncoding.EncodeToString(bytes) != signature {
		return nil, receiptError()
	}
	if err = core.VerifyCertificateData(owner, data, bytes); err != nil {
		return nil, err
	}
	if _, err = bounded(cert, ContributionReceiptBytes); err != nil {
		return nil, err
	}
	return resourceClone(cert)
}

// A signing-owned journal must authorize this internal operation; no RPC exposes it.
func CreateContributionReceipt(identity core.Identity, input any) (map[string]any, error) {
	q, err := object(input, receiptFields...)
	if err != nil {
		return nil, receiptError()
	}
	body := map[string]any{"domain": "relayloom/site-contribution-receipt/1", "owner": identity.Public}
	for k, v := range q {
		body[k] = v
	}
	if _, _, err = receiptBody(body); err != nil {
		return nil, err
	}
	data, err := bounded(body, ContributionReceiptBytes)
	if err != nil {
		return nil, err
	}
	signature, err := core.SignCertificateData(identity, data)
	if err != nil {
		return nil, err
	}
	return VerifyContributionReceipt(map[string]any{"body": body, "id": core.Hash(data), "signature": base64.StdEncoding.EncodeToString(signature)})
}
func MatchContributionReceipt(value, proposalValue any) (map[string]any, error) {
	receipt, err := VerifyContributionReceipt(value)
	if err != nil {
		return nil, err
	}
	proposal, err := VerifyContribution(proposalValue)
	if err != nil {
		return nil, err
	}
	r, b := receipt["body"].(map[string]any), proposal["body"].(map[string]any)
	if r["certificateId"] != proposal["id"] || r["contributorId"] != b["contributor"].(map[string]any)["id"] || r["operationId"] != b["operationId"] || !creationEqual(r["target"], b["target"]) || !creationEqual(r["proposalCreated"], b["created"]) || !creationEqual(r["proposalExpires"], b["expires"]) {
		return nil, receiptError()
	}
	return receipt, nil
}
func ParseContributionReceiptContent(input any) (map[string]any, error) {
	m, err := object(input, "type", "receipt")
	if err != nil || m["type"] != "site-contribution-receipt" {
		return nil, receiptError()
	}
	cert, err := VerifyContributionReceipt(m["receipt"])
	if err != nil {
		return nil, err
	}
	return map[string]any{"type": "site-contribution-receipt", "receipt": cert}, nil
}

// Caller must authenticate/decrypt the outer bundle and enforce live policy.
func MatchContributionReceiptEnvelope(bundle core.Bundle, plaintext any) (map[string]any, error) {
	content, err := ParseContributionReceiptContent(plaintext)
	if err != nil {
		return nil, err
	}
	b := content["receipt"].(map[string]any)["body"].(map[string]any)
	owner := b["owner"].(map[string]any)
	expected := []string{docTextValue(owner["id"])}
	if b["contributorId"] != owner["id"] {
		expected = append(expected, docTextValue(b["contributorId"]))
	}
	sort.Strings(expected)
	created, _ := contributionClock(b["created"])
	expires, _ := contributionClock(b["expires"])
	m := bundle.Manifest
	if m.Kind != "site-contribution-receipt" || !creationEqual(m.Author, owner) || m.PublicKey != nil || !creationEqual(readersOf(bundle), expected) || m.Created != created || m.Expires != expires {
		return nil, receiptError()
	}
	return content, nil
}

File: packages/sites/src/contribution-receipt.ts
import {
  canonical,
  exactShape,
  type Bundle,
  type Identity,
  type PublicIdentity,
} from "../../core/src/protocol";
import type { CertificateCrypto } from "../../core/src/certificate-types";
import { parseSiteAddress } from "./protocol";
import {
  createSiteContributionProtocol,
  SITE_CONTRIBUTION_LIMITS,
  type ContributionTarget,
} from "./contribution-protocol";

export const CONTRIBUTION_RECEIPT_LIMITS = Object.freeze({
  bytes: 8192,
  lifetimeMs: 30 * 86400_000,
});
export interface ContributionReceiptRequest {
  contributorId: string;
  certificateId: string;
  operationId: string;
  target: ContributionTarget;
  proposalCreated: number;
  proposalExpires: number;
  verifiedAt: number;
  created: number;
  expires: number;
}
export interface ContributionReceiptBody extends ContributionReceiptRequest {
  domain: "relayloom/site-contribution-receipt/1";
  owner: PublicIdentity;
}
export interface ContributionReceipt {
  body: ContributionReceiptBody;
  id: string;
  signature: string;
}
export interface ContributionReceiptContent {
  type: "site-contribution-receipt";
  receipt: ContributionReceipt;
}
const fields = [
  "contributorId",
  "certificateId",
  "operationId",
  "target",
  "proposalCreated",
  "proposalExpires",
  "verifiedAt",
  "created",
  "expires",
];
const id = (v: unknown): v is string =>
  typeof v === "string" && /^[a-f0-9]{64}$/.test(v);
const clock = (v: unknown): v is number =>
  Number.isSafeInteger(v) && (v as number) >= 0;
function insist(ok: unknown, reason: string): asserts ok {
  if (!ok) throw Error("Recibo de proposta inválido: " + reason);
}

/** An owner's attestation of historical verified reception. This is never a
 * publication decision, visitor grant, transport ACK or proof of clock accuracy.
 * Only signing-owned catalogues may derive an intent from authenticated evidence;
 * none of these signing/intent helpers is an RPC operation. */
export function createContributionReceiptProtocol(crypto: CertificateCrypto) {
  const proposals = createSiteContributionProtocol(crypto);
  function bounded(value: unknown) {
    const text = canonical(value);
    insist(
      new TextEncoder().encode(text).length <=
        CONTRIBUTION_RECEIPT_LIMITS.bytes,
      "orçamento",
    );
    return text;
  }
  function body(input: unknown): asserts input is ContributionReceiptBody {
    insist(exactShape(input, ["domain", "owner", ...fields]), "campos");
    const b = input as ContributionReceiptBody;
    insist(
      b.domain === "relayloom/site-contribution-receipt/1" &&
        crypto.validateIdentity(b.owner),
      "dono ou domínio",
    );
    insist(
      id(b.contributorId) &&
        id(b.certificateId) &&
        typeof b.operationId === "string" &&
        /^[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89ab][a-f0-9]{3}-[a-f0-9]{12}$/.test(
          b.operationId,
        ),
      "referência da proposta",
    );
    insist(
      exactShape(b.target, [
        "site",
        "snapshotId",
        "revisionId",
        "pageId",
        "formId",
      ]),
      "destino",
    );
    insist(
      parseSiteAddress(b.target.site).ownerId === b.owner.id &&
        id(b.target.snapshotId) &&
        id(b.target.revisionId) &&
        [b.target.pageId, b.target.formId].every(
          (v) => typeof v === "string" && /^[A-Za-z0-9_-]{1,64}$/.test(v),
        ),
      "dono ou referência do destino",
    );
    insist(
      [
        b.proposalCreated,
        b.proposalExpires,
        b.verifiedAt,
        b.created,
        b.expires,
      ].every(clock) &&
        b.proposalExpires > b.proposalCreated &&
        b.proposalExpires - b.proposalCreated <=
          SITE_CONTRIBUTION_LIMITS.lifetimeMs &&
        b.verifiedAt < b.proposalExpires &&
        b.proposalCreated - b.verifiedAt <=
          SITE_CONTRIBUTION_LIMITS.clockSkewMs &&
        b.created === b.verifiedAt &&
        b.expires > b.created &&
        b.expires - b.created <= CONTRIBUTION_RECEIPT_LIMITS.lifetimeMs,
      "prazo histórico ou intenção renovada",
    );
    bounded(b);
  }
  function verify(input: unknown): ContributionReceipt {
    insist(exactShape(input, ["body", "id", "signature"]), "certificado");
    const value = input as ContributionReceipt;
    body(value.body);
    insist(
      id(value.id) &&
        typeof value.signature === "string" &&
        value.signature.length === 88,
      "assinatura",
    );
    const bytes = Uint8Array.from(atob(value.signature), (c) =>
      c.charCodeAt(0),
    );
    insist(
      bytes.length === 64 &&
        btoa(String.fromCharCode(...bytes)) === value.signature,
      "assinatura não canónica",
    );
    const text = bounded(value.body);
    insist(
      value.id === crypto.hash(text) &&
        crypto.verify(value.body.owner, text, bytes),
      "assinatura ou hash",
    );
    return JSON.parse(bounded(value));
  }
  function request(
    input: unknown,
    owner: PublicIdentity,
  ): ContributionReceiptRequest {
    insist(exactShape(input, fields), "intenção");
    body({
      domain: "relayloom/site-contribution-receipt/1",
      owner,
      ...(input as ContributionReceiptRequest),
    });
    return JSON.parse(bounded(input));
  }
  function create(
    identity: Identity,
    request: ContributionReceiptRequest,
  ): ContributionReceipt {
    insist(exactShape(request, fields), "intenção");
    const value: ContributionReceiptBody = {
      domain: "relayloom/site-contribution-receipt/1",
      owner: identity.public,
      ...request,
    };
    body(value);
    const owned = JSON.parse(bounded(value)),
      text = bounded(owned);
    return verify({
      body: owned,
      id: crypto.hash(text),
      signature: crypto.sign(identity, text),
    });
  }
  /** Historical binding; does not check admission or create publication authority. */
  function matchProposal(input: unknown, proposalInput: unknown) {
    const receipt = verify(input),
      proposal = proposals.verify(proposalInput),
      r = receipt.body,
      b = proposal.body;
    insist(
      r.certificateId === proposal.id &&
        r.contributorId === b.contributor.id &&
        r.operationId === b.operationId &&
        canonical(r.target) === canonical(b.target) &&
        r.proposalCreated === b.created &&
        r.proposalExpires === b.expires,
      "recibo de outra proposta",
    );
    return receipt;
  }
  function content(input: unknown): ContributionReceiptContent {
    insist(
      exactShape(input, ["type", "receipt"]) &&
        (input as ContributionReceiptContent).type ===
          "site-contribution-receipt",
      "conteúdo",
    );
    return {
      type: "site-contribution-receipt",
      receipt: verify((input as ContributionReceiptContent).receipt),
    };
  }
  /** Caller must authenticate/decrypt the outer bundle and enforce live policy. */
  function matchEnvelope(bundle: Bundle, plaintext: unknown) {
    const value = content(plaintext),
      b = value.receipt.body,
      m = bundle.manifest,
      readers = [...new Set([b.owner.id, b.contributorId])].sort();
    insist(
      m.kind === "site-contribution-receipt" &&
        canonical(m.author) === canonical(b.owner) &&
        m.publicKey === null &&
        canonical(m.keys.map((k) => k.reader).sort()) === canonical(readers) &&
        m.created === b.created &&
        m.expires === b.expires,
      "envelope diferente do recibo privado",
    );
    return value;
  }
  return { create, verify, matchProposal, content, matchEnvelope, request };
}
